transformer_lens/components/transformer_block.py

```diff
@@ -157,21 +168,33 @@ def forward(
                 attention_mask=attention_mask,
             )
         )  # [batch, pos, d_model]
+        if self.cfg.use_normalization_before_and_after:
+            # If we use LayerNorm both before and after, then apply the second LN after the layer
+            # and before the hook. We do it before the hook so hook_attn_out captures "that which
+            # is added to the residual stream"
+            attn_out = self.ln1_post(attn_out)
+        attn_out = self.hook_attn_out(attn_out)
         if not self.cfg.attn_only and not self.cfg.parallel_attn_mlp:
             resid_mid = self.hook_resid_mid(resid_pre + attn_out)  # [batch, pos, d_model]
             mlp_in = (
                 resid_mid if not self.cfg.use_hook_mlp_in else self.hook_mlp_in(resid_mid.clone())
             )
             normalized_resid_mid = self.ln2(mlp_in)
-            mlp_out = self.hook_mlp_out(self.mlp(normalized_resid_mid))  # [batch, pos, d_model]
+            mlp_out = self.mlp(normalized_resid_mid)  # [batch, pos, d_model]
+            if self.cfg.use_normalization_before_and_after:
+                mlp_out = self.ln2_post(mlp_out)
+            mlp_out = self.hook_mlp_out(mlp_out)
```

Comment:
Can we break this out into a function? Not a fan of repeated logic, as it is often where bugs can happen either through things like initial typos, or someone changing something down the line without realizing the logic exists twice. Maybe a function named `apply_mlp` that we can call here, and then move this block into that.

Reply:
The repeated logic being that this is done after attn _and_ after mlp? I don't see a good way to combine it in abstract logic, since they use different LNs (I could make a 'apply_normalisation_after' function that takes as input the output and whether it's attn or mlp, and chooses the right LN accordingly, but that's kinda ugly and hard to read IMO.

We could pull it out of the forward pass into a separate fn, but I don't think this would do much to minimise risk if there's a separate apply_mlp and apply_attn function.

Reply:
I am referring to this line, and line 194. They are currently identical outside of the input variable.

Reply:
Oh, gotcha. Yeah that seems reasonable to me